### PR TITLE
Add section on Python versions

### PR DESCRIPTION
**Pull request by samaocarpenter**
This adds a short section on Python versions, how to interpret version numbers, and what updates mean.

It also corrects my GitHub link in the intro page, since I changed my URL a couple months ago

### CODE REVIEW

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 1]
```diff
@@ -144,6 +144,39 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
 We will be relying heavily on Python and a library for doing optimized numerical computations, called NumPy, throughout this course.
 <!-- #endregion -->
 
+<!-- #region -->
+## Keeping Up To Date
+
+New versions of Python come out periodically, bringing new features and fixes. 
+The most recent version of Python is Python 3.8.4.
```
Comment (davidmascharka) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:151]: I'd generally advise against these kinds of statements because it's going to be either (1) a pain to update every time a new release comes out or (2) just be outdated

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 2]
```diff
@@ -144,6 +144,39 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+Unfortunately, it takes time for packages to support new versions of Python.
+As a result, many packages don't support Python 3.8, and will require you to install Python 3.7 or, in some cases, 3.6.
```
Comment (davidmascharka) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:153]: I think this could be generalized and just say that version incompatibilities will sometimes crop up :shrug:

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 3]
```diff
@@ -144,6 +144,39 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+
+All Python versions numbers use the A.B.C format.
```
Comment (davidmascharka) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:155]:
```suggestion
All Python version numbers use the A.B.C format. 
```

Further, they conform to [semantic versioning](https://semver.org/) - do we want to include a link?

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 4]
```diff
@@ -144,6 +144,39 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+
+All Python versions numbers use the A.B.C format. 
+The three numbers denote major releases, minor releases, and patches.
+
+- The first number denotes major releases to the language. 
+The most current major release is Python 3, but Python 2 is still in use. 
+When a major release comes out, it means that older code will not necessarily work with the new release. 
+Running Python 2 code in a Python 3 interpreter can yield errors, as can running Python 3 code in a Python 2 interpreter. 
+Python 2 is no longer supported, but the transition to Python 3 has been slow.
+When reading code from the Internet, always check to make sure it uses Python 3.
```
Comment (davidmascharka) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:163]: I'd rearrange the order here

```suggestion
- The first number denotes major releases to the language. 
When a major release comes out, it means that older code will not necessarily work with the new release. 
The most current major release is Python 3. Running Python 2 code in a Python 3 interpreter can yield errors, as can running Python 3 code in a Python 2 interpreter. 
Python 2 is no longer supported, but the transition to Python 3 has been slow, so Python 2 is still in use in some places.
When reading code from the Internet, always check to make sure it uses Python 3.
```

I'm not sure we want to talk much about 2/3 since 2 is officially dead... @rsokl?

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 5]
```diff
@@ -144,6 +144,39 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+When a minor release comes out, older code will run in the new interpreter, but new code will not necessarily be compatible with older versions. 
+Python 3.7 and 3.8 are great examples of this - any code that works in Python 3.7 will run in Python 3.8, but new syntax has been added that will not work in Python 3.7.
+
+- The third and final number denotes a patch, which generally means bug fixes and performance improvements. 
+All code within the same minor release will run on all other patches within that minor release - all Python 3.7.8 code is compatible with a Python 3.7.1 interpreter, and vice versa. 
+Patches are released fairly often, and their changes only occur 'under the hood'.
+
+We will be using Python 3 in this course. 
+Python 3.7.8 is compatible with most packages while Python 3.8.4 is not yet, so if you need to use a variety of packages it is safer to stick with Python 3.7.8. 
+Otherwise, you will have no problems switching to Python 3.8.4.
```
Comment (davidmascharka) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:175]: For the same reason as above, I'd remove the patch version from this paragraph and just use 3.7/3.8 or these will be outdated quickly

Comment (rsokl) [reply Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:175]: Yeah, I am hesitant to refer to specific versions of Python at all - unless we are speaking to specific features or deprecations

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md (hunk 4 above)
Comment (rsokl) [reply Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:163]: It might be okay to talk about Python 2 here since it is the one handy example of switching between major versions. That being said, the convo about it should be brief.

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md (hunk 3 above)
Comment (rsokl) [reply Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:155]: I think a link would be good

**Comment by samaocarpenter** [comment]
I addressed the comments you guys made - thank you!

I tried to roll back references to specific versions, but some have remained as examples. Let me know if you think there's still too much specific versioning

**Comment by samaocarpenter** [comment]
I made some minor edits just to add some clarity

**Comment by rsokl** [comment]
Will review and merge this ASAP!

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 6]
```diff
@@ -144,6 +144,42 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
 
+<!-- #region -->
+## Keeping Up To Date
+
+New versions of Python come out periodically, bringing new features and fixes. 
+It's important to keep Python updated in order to make sure that you have access to the latest additions to the language.
+Unfortunately, it takes time for packages to support the newest versions of Python.
+As a result, it can sometimes be necessary to use older versions to maintain compatibility.
+
+All Python version numbers use the A.B.C format, in accordance with [semantic versioning](https://semver.org/).
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:155]:
```suggestion
All Python version numbers use the `A.B.C` format, in accordance with [semantic versioning](https://semver.org/). 
```

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 7]
```diff
@@ -144,6 +144,42 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+
+New versions of Python come out periodically, bringing new features and fixes. 
+It's important to keep Python updated in order to make sure that you have access to the latest additions to the language.
+Unfortunately, it takes time for packages to support the newest versions of Python.
+As a result, it can sometimes be necessary to use older versions to maintain compatibility.
+
+All Python version numbers use the A.B.C format, in accordance with [semantic versioning](https://semver.org/). 
+The three numbers denote major releases, minor releases, and patches.
+
+- The first number denotes major releases to the language.
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:158]:
```suggestion
The first number denotes major releases to the language. 
```

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 8]
```diff
@@ -144,6 +144,42 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+The three numbers denote major releases, minor releases, and patches.
+
+- The first number denotes major releases to the language. 
+When a major release comes out, it means that older code will not necessarily work with the new release, and vice versa. 
+The most current major release is Python 3, but Python 2 is still occasionally used. 
+Running Python 2 code in a Python 3 interpreter can yield errors, as can running Python 3 code in a Python 2 interpreter. 
+Python 2 has been deprecated and should not be used, but the transition to Python 3 has been slow.
+When reading code from the Internet, always check to make sure it uses Python 3.
+
+- The second number denotes a minor release.
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:165]:
```suggestion
The second number denotes a minor release. 
```

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 9]
```diff
@@ -144,6 +144,42 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+The most current major release is Python 3, but Python 2 is still occasionally used. 
+Running Python 2 code in a Python 3 interpreter can yield errors, as can running Python 3 code in a Python 2 interpreter. 
+Python 2 has been deprecated and should not be used, but the transition to Python 3 has been slow.
+When reading code from the Internet, always check to make sure it uses Python 3.
+
+- The second number denotes a minor release. 
+When a minor release comes out, older code will run in the new interpreter, but new code will not necessarily be compatible with older versions.
+For example, any code that works in Python 3.7 will run in Python 3.8, but because Python 3.8 added new syntax, code from Python 3.8 will not necessarily run in a Python 3.7 interpreter.
+
+- The third and final number denotes a patch, which generally means bug fixes and performance improvements.
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:169]:
```suggestion
The third and final number denotes a patch, which generally means bug fixes and performance improvements. 
```

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 10]
```diff
@@ -144,6 +144,42 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+It's important to keep Python updated in order to make sure that you have access to the latest additions to the language.
+Unfortunately, it takes time for packages to support the newest versions of Python.
+As a result, it can sometimes be necessary to use older versions to maintain compatibility.
+
+All Python version numbers use the A.B.C format, in accordance with [semantic versioning](https://semver.org/). 
+The three numbers denote major releases, minor releases, and patches.
+
+- The first number denotes major releases to the language. 
+When a major release comes out, it means that older code will not necessarily work with the new release, and vice versa. 
+The most current major release is Python 3, but Python 2 is still occasionally used.
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:160]:
```suggestion
The most current major release is Python 3; Python 2 is no longer supported by any bug or security fixes.
All releases in the near future will be improvements to Python 3, and thus they will come in the form of minor releases and patches.
```

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 11]
```diff
@@ -144,6 +144,42 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+
+All Python version numbers use the A.B.C format, in accordance with [semantic versioning](https://semver.org/). 
+The three numbers denote major releases, minor releases, and patches.
+
+- The first number denotes major releases to the language. 
+When a major release comes out, it means that older code will not necessarily work with the new release, and vice versa. 
+The most current major release is Python 3, but Python 2 is still occasionally used. 
+Running Python 2 code in a Python 3 interpreter can yield errors, as can running Python 3 code in a Python 2 interpreter. 
+Python 2 has been deprecated and should not be used, but the transition to Python 3 has been slow.
+When reading code from the Internet, always check to make sure it uses Python 3.
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:163]:
```suggestion
```

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 12]
```diff
@@ -144,6 +144,42 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+- The second number denotes a minor release. 
+When a minor release comes out, older code will run in the new interpreter, but new code will not necessarily be compatible with older versions.
+For example, any code that works in Python 3.7 will run in Python 3.8, but because Python 3.8 added new syntax, code from Python 3.8 will not necessarily run in a Python 3.7 interpreter.
+
+- The third and final number denotes a patch, which generally means bug fixes and performance improvements. 
+All code within the same minor release will run on all other patches within that minor release - for example, all Python 3.7.8 code is compatible with a Python 3.7.1 interpreter, and vice versa. 
+Patches are released fairly often, and their changes only occur 'under the hood'.
+
+In simpler terms, major releases are neither backward nor forward compatible.
+Minor releases are backward compatible but not forward compatible, and patches are both forward and backward compatible.
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:174]: New features / changes in minor releases may not be backwards compatible.

```suggestion
Minor releases are forward compatible but not necessarily fully backward compatible, and patches are both forward and backward compatible.
```

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 13]
```diff
@@ -144,6 +144,42 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+
+We will be using Python 3 in this course. 
+If you use a wide variety of packages, you may need to use an older minor release of Python 3 to ensure compatibility.
+Otherwise, you should use whatever [the most current version of Python](https://www.python.org/downloads/) is.
+Luckily, Anaconda manages compatibility for you, so you shouldn't have to worry too much about which version of Python you use as long as you occasionally check for updates.
+
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:180]: It is stated earlier that Python 2 is no longer supported.

```suggestion
```

**Review by rsokl — CHANGES_REQUESTED** [review]

**Comment by samaocarpenter** [comment]
@rsokl thanks for the review - I addressed the comments you made, and made a couple small changes to enhance the clarity

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md [hunk 14]
```diff
@@ -144,6 +144,37 @@ As such, Python is a language that is conducive to rapidly prototyping and testi
...
+All code within the same minor release will run on all other patches within that minor release - for example, all Python 3.7.8 code is compatible with a Python 3.7.1 interpreter, and vice versa. 
+Patches are released fairly often, and their changes only occur 'under the hood'.
+
+In simpler terms, major releases are neither backward nor forward compatible.
+Minor releases are forward compatible but not necessarily fully backward compatible, and patches are both forward and backward compatible.
+
+Unfortunately, it takes time for packages to support the newest versions of Python.
+As a result, it can sometimes be necessary to use older versions to maintain compatibility.
+Anaconda will manage package compatability for you most of the time, but be careful not to mix incompatible libraries.
+
```
Comment (rsokl) [inline Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md:175]: I'd prefer to remove this section, because it doesn't really tell people how to avoid mixing incompatible libraries, and, in fact, the conda package manager should in principle always handle this

```suggestion
```

**Review by rsokl — CHANGES_REQUESTED** [review]

File: Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md (hunk 14 above)
Comment (samaocarpenter) [reply Python/Module1_GettingStartedWithPython/GettingStartedWithPython.md]: That makes sense - just committed the change

**Review by rsokl — APPROVED** [review]
